lib/elixir/lib/map.ex

```diff
@@ -211,13 +220,16 @@ defmodule Map do
   """
   @spec take(map, [key]) :: map
   def take(map, keys) do
-    Enum.reduce(keys, [], fn key, acc ->
-      case fetch(map, key) do
-        {:ok, value} -> [{key, value} | acc]
-        :error -> acc
-      end
-    end)
-    |> :maps.from_list
+    do_take(keys, map, [])
+  end
```

Comment:
With all that said, I am not sure this optimization is worth making. `Enum.reduce` is supposed to be quite fast in both split, take and drop cases. If we want to optimize, we could inline only when the keys are a list.